src/utils.py

```diff
@@ -0,0 +1,54 @@
+import os
+import re
+from semver import Version
+
+
+def get_required_packages_from_env_in(file_path) -> set[str]:
+    required_packages = set()
+    if not os.path.exists(file_path):
+        return required_packages
+    with open(file_path, 'r') as f:
+        lines = f.readlines()
+        required_packages = {line.strip().split("::")[1].split("[")[0] for line in lines if line.startswith('conda-forge::')}
+    return required_packages
+
+
+def get_installed_packages_and_conda_forge_sub_dir_from_env_out(file_path) -> (dict[str, str],
+                                                                               dict[str, str]):
+    installed_packages = dict()
+    package_to_subdirectory_mapping = dict()
+    if not os.path.exists(file_path):
+        return installed_packages, package_to_subdirectory_mapping
+    with open(file_path, 'r') as f:
```

Comment:
Same comment as earlier - please use official Conda APIs.

I suspect a single API might work to read both *env.in* and *env.out* files.

Reply:
But ultimately in this case we are reading env.out file and parse the download url of each marquee package.

Assume that we aren't using conda forge, this env.out will have a different url format corresponding to the channel that was used.

Conda doesn't dictate how a channel organizes its packages. Hence we need to come up with our regex for conda forge urls

Reply:
The main task in this method is to parse a conda forge url. 
Don't think conda has specific methods for parsing conda forge url